snuba/utils/streams/consumers/backends/synchronized.py

```diff
@@ -0,0 +1,271 @@
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from dataclasses import dataclass, field
+from threading import Event, Lock
+from typing import Callable, Generic, Mapping, MutableMapping, Optional, Set, Sequence
+
+from snuba.utils.concurrent import execute
+from snuba.utils.streams.consumers.backends.abstract import ConsumerBackend
+from snuba.utils.streams.consumers.types import Message, TOffset, TStream, TValue
+
+
+ConsumerGroup = str
+
+
+@dataclass
+class CommitData(Generic[TStream, TOffset]):
+    consumer_group: ConsumerGroup
+    stream: TStream
+    offset: TOffset
+
+
+# TODO: This should be moved into the abstract consumer backend.
+class Topic(ABC, Generic[TStream]):
+    @abstractmethod
+    def __contains__(self, stream: TStream) -> bool:
+        raise NotImplementedError
+
+
+@dataclass
+class Subscription(Generic[TStream, TOffset]):
+    topics: Sequence[Topic[TStream]]
+    remote_consumer_group_offsets: MutableMapping[
+        TStream, MutableMapping[ConsumerGroup, TOffset]
+    ] = field(default_factory=lambda: defaultdict(dict))
+
+
+# TODO: TOffset needs to be bound as some sort of comparable type.
```

Comment:
Not sure what the best approach for this will be. Right now, mypy rightfully complains about these comparisons:

```
snuba/utils/streams/consumers/backends/synchronized.py:179: error: Unsupported left operand type for >= ("TOffset")
snuba/utils/streams/consumers/backends/synchronized.py:222: error: Unsupported left operand type for > ("TOffset")
snuba/utils/streams/consumers/backends/synchronized.py:238: error: Unsupported left operand type for >= ("TOffset")
```

Reply:
See 12eca8a: this introduces a `Comparable` protocol that is used as type bound, which ensures that `TOffset` values have [a total ordering](https://docs.python.org/3.7/library/functools.html#functools.total_ordering).